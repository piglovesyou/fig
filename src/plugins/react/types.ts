import { NodePath } from '@babel/traverse';
import { JSXElement } from '@babel/types';

export type ReactCursorType = NodePath<JSXElement>;
