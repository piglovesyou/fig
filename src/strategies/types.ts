import { NodePath } from '@babel/traverse';
import { JSXElement } from '@babel/types';
import { GenContext } from '../gen/make-gen-context';
import {
  ParentVisitContext,
  VisitContext,
  VisitContextWithCursor,
} from '../visit/visit-context';

export interface Strategy {
  makeLayout(): NodePath<JSXElement>;
  postWalk(): void;
  render(): string;
  appendComponentInstanceElement(
    context: VisitContext,
    parentContext: VisitContextWithCursor,
    genContext: GenContext
  ): void;
  appendElement(
    context: VisitContext,
    parentContext: VisitContextWithCursor,
    tagName?: string
  ): NodePath<JSXElement>;
  appendSvgElement(
    context: VisitContext,
    parentContext: ParentVisitContext,
    svgHtml: string
  ): void;
  appendTextElement(
    context: VisitContext,
    parentContext: ParentVisitContext
  ): void;
}
