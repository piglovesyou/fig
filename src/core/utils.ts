import { ComposableNode, Node } from '../types/ast';

export function isValidComponentNode(
  node: ComposableNode
): node is Node<'FRAME'> | Node<'COMPONENT'> | Node<'INSTANCE'> {
  return (
    (node.type === 'FRAME' ||
      node.type === 'COMPONENT' ||
      node.type === 'INSTANCE') &&
    node.visible !== false
  );
}

export function walkNodeTree<T = void>(
  node: ComposableNode,
  fn: (node: ComposableNode, parentResult: T) => null | T,
  parentResult: T
): void {
  const result = fn(node, parentResult);
  if (result === null) return;
  for (const child of node.children || []) {
    walkNodeTree(child, fn, result);
  }
}
