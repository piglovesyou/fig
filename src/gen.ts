import { writeFile } from 'fs/promises';
import { requestFile } from './api';
import { makeGenContext } from './make-gen-context';
import { processCanvas } from './process-canvas';
import { Canvas, FigmaFile } from './types';

export async function gen(outDir: string, fileKey: string) {
  const figmaFile: FigmaFile = await requestFile(fileKey);

  // TODO: for dev
  await writeFile(
    `${__dirname}/__fixtures/${fileKey}.file.json`,
    JSON.stringify(figmaFile, undefined, 2)
  );
  const genContext = await makeGenContext(figmaFile, fileKey, outDir);

  for (const canvas of figmaFile.document.children as Canvas[])
    await processCanvas(canvas, genContext);
}
