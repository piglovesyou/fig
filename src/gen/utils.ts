import camelCase from 'camelcase';
import { ComposableNode } from '../types/ast';

function makeComponentNameFromId(id: string, name: string) {
  id = id.replace(/;/g, '$$').replace(/\W/g, '$');
  name = camelCase(name, { pascalCase: true });
  return `${name}_${id}`;
}

export function makeComponentName(node: ComposableNode): string | never {
  switch (node.type) {
    case 'FRAME':
    case 'COMPONENT':
      return makeComponentNameFromId(node.id, node.name);
    case 'INSTANCE':
      // TODO: zombie component name can be resolved. Fix it.
      return makeComponentNameFromId(node.id, '__NOT_FOUND__');
    default:
      throw new Error('Component can only be FRAME, COMPONENT, INSTANCE.');
  }
}
