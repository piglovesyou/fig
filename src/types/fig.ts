export * from './ast';
import { ComposableNode, Node as _Node } from './ast';

export interface FigmaFile {
  document: Document;
  components: Record<string, ComponentRef>;
  schemaVersion: number;
  styles: Styles;
  name: string;
  lastModified: string;
  thumbnailUrl: string;
  version: string;
  role: string;
}

export type Document = _Node<'DOCUMENT'> & {
  children: Canvas[];
};

export type Canvas = _Node<'CANVAS'> & {
  children: ComposableNode[];
};

// type MetaProps = {
//   asVector?: boolean;
//   // children?: PrimitiveNode[];
// };

// export type ComposableNode =
//   MetaProps & {
//     children: ComposableNode;
//   };

// DOCUMENT
// CANVAS
// FRAME
// GROUP
// VECTOR
// BOOLEAN
// BOOLEAN_OPERATION
// STAR
// LINE
// ELLIPSE
// REGULAR_POLYGON
// RECTANGLE
// TEXT
// SLICE
// COMPONENT
// INSTANCE

export interface PrototypeDevice {
  type: string;
  rotation: string;
}

export interface ComponentRef {
  key: string;
  name: string;
  description: string;
}

export interface Styles {}

export interface Bound {
  top: number;
  right: number;
  bottom: number;
  left: number;
  width: number;
  height: number;
}
