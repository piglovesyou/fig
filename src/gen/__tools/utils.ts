import generate from '@babel/generator';
import { NodePath } from '@babel/traverse';
import { JSXElement } from '@babel/types';
import { format } from 'prettier';
import { createConfig } from '../../core/config';
import { ComposableNode, FigmaFile } from '../../types/fig';
import { readFile } from '../../utils/fs';
import { makeGenContext } from '../make-gen-context';

export async function makeCreateComponentArgs(
  fileJsonFullPath: string,
  fileKey: string,
  baseDir: string
) {
  const figmaFileFixture = JSON.parse(
    await readFile(fileJsonFullPath, 'utf-8')
  );
  const node: ComposableNode =
    figmaFileFixture.document.children[0].children[0];
  expect(node.name).toBe('My Frame');

  const config = await createConfig({
    fileKeys: [fileKey],
    baseDir,
    plugins: ['react'],
  });
  const genContext = await makeGenContext(
    figmaFileFixture as FigmaFile,
    fileKey,
    config,
    baseDir
  );

  const { componentsMap } = genContext;
  const componentInfo = componentsMap.get('1:2');
  if (!componentInfo) throw new Error('Never');
  return [node, componentInfo, genContext] as const;
}

export function getCode(rootCursor: NodePath<JSXElement>) {
  const { code } = generate(rootCursor.node);
  return format(code, { parser: 'babel' });
}
