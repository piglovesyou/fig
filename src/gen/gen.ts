import makeDir from 'make-dir';
import pMap from 'p-map';
import pReduce from 'p-reduce';
import { join } from 'path';
import { requestFile } from '../core/api';
import { FigConfig } from '../core/config';
import { updateLog, updateLogDone } from '../core/print';
import { FigmaFile } from '../types/fig';
import { ComponentInfo, GenContext } from '../types/gen';
import { FigPlugin } from '../types/plugin';
import { writeFile } from '../utils/fs';
import { appendImagesMap, makeExistingImagesMap } from './images-map';
import { makeGenContext, makePaths } from './make-gen-context';
import { processComponent } from './process-component';
import { appendVectorsMap } from './vectors-map';

export async function processHtml(
  componentInfo: ComponentInfo,
  genContext: GenContext
) {
  const { node, name } = componentInfo;
  if (node.type !== 'FRAME')
    throw new Error('Never. node.type should be FRAME');
  const { plugins } = genContext;
  if (!plugins) throw new Error('Never. Plugins should be instantiated.');

  const html = (await pReduce(
    plugins,
    (html, plugin) => {
      return plugin.renderHtml(componentInfo, genContext);
    },
    null as null | ReturnType<FigPlugin<unknown>['renderHtml']>
  ))!;

  await writeFile(join(genContext.htmlFullDir, name + '.html'), html);
}

export async function gen(
  config: FigConfig,
  cwd = process.cwd()
): Promise<void> {
  const { fileKeys, token } = config;
  for (const fileKey of fileKeys) {
    updateLog(`Fetching data of "${fileKey}"..`);
    const figmaFile: FigmaFile = await requestFile(fileKey, token);

    updateLog(`Fetching image resources..`);
    const genContext = await makeGenContext(figmaFile, fileKey, config, cwd);
    const { componentsMap, plugins } = genContext;
    if (!plugins) throw new Error('Never. Plugins should be instantiated.');

    const existingImagesMap = await makeExistingImagesMap(
      genContext.imagesFullDir
    );
    const paths = makePaths(config); // refactor
    await appendImagesMap(
      genContext.imagesMap,
      paths,
      fileKey,
      token,
      existingImagesMap
    );
    await appendVectorsMap(
      paths,
      genContext.vectorsMap,
      genContext.vectorsList,
      fileKey,
      token,
      existingImagesMap
    );

    const components: ComponentInfo[] = [];
    const frames: ComponentInfo[] = [];
    for (const [, componentInfo] of componentsMap) {
      const { node } = componentInfo;
      if (node.type === 'FRAME') frames.push(componentInfo);
      else components.push(componentInfo);
    }

    // Generate components to "./components"
    updateLog(`Generating ${components.length} components..`);
    await pMap(componentsMap, ([, componentInfo]) =>
      processComponent(componentInfo, genContext)
    );

    // Generate components to "./pages"
    updateLog(`Generating ${frames.length} page components..`);
    await pMap(componentsMap, ([, componentInfo]) =>
      processComponent(componentInfo, genContext)
    );

    // Generate html to "./public"
    updateLog(`Generating ${frames.length} HTMLs..`);
    await makeDir(genContext.htmlFullDir);
    await pMap(frames, (componentInfo) => {
      return processHtml(componentInfo, genContext);
    });

    for (const plugin of plugins) await plugin.dispose();

    figmaFile.name;
    updateLog(
      `"${figmaFile.name}" done. ${genContext.imagesMap.size} images, ${components.length} components, ${frames.length} pages and HTMLs are synchronized.`
    );
    updateLogDone();
  }
}
