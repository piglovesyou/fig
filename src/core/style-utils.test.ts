import { makeComponentName } from '../gen/utils';
import { ComposableNode } from '../types/fig';

describe('figma.test.ts', () => {
  test('makeComponentName()', async () => {
    const rv = makeComponentName({
      id: '1:1',
      name: 'a',
      type: 'COMPONENT',
    } as ComposableNode);
    expect(rv).toMatchSnapshot();
  });
});
