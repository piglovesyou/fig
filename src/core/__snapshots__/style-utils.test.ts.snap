// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`figma.test.ts makeComponentName() 1`] = `"A_1$1"`;
